cumulusci/oauth/client.py

```diff
@@ -0,0 +1,293 @@
+            # https://github.com/SFDO-Tooling/CumulusCI/pull/2373
+            old_timeout = socket.getdefaulttimeout()
+            socket.setdefaulttimeout(3)
+            self.httpd.serve_forever()
+        finally:
+            socket.setdefaulttimeout(old_timeout)
+
+        if use_https:
+            Path("key.pem").unlink()
+            Path("localhost.pem").unlink()
+
+        # timeout thread can stop polling and just finish
+        timeout_thread.quit()
+        self.validate_response(self.response)
+        return safe_json_from_response(self.response)
+
+    def _get_auth_uri(self):
+        url = self.client_info.auth_uri
+        url += "?response_type=code"
+        url += f"&client_id={self.client_info.client_id}"
+        url += f"&redirect_uri={self.client_info.callback_url}"
+        if self.client_info.scope:
+            url += f"&scope={quote(self.client_info.scope)}"
+        if self.client_info.prompt:
+            url += f"&prompt={quote('login')}"
+        return url
+
+    def _create_httpd(self, use_https=False):
+        """Create an http deamon process to listen
+        for the callback from the auth server"""
+        print(f"_create_httpd(use_https={use_https})")
+        url_parts = urlparse(self.client_info.callback_url)
+        server_address = (url_parts.hostname, url_parts.port)
+        OAuthCallbackHandler.parent = self
+
+        httpd = HTTPServer(server_address, OAuthCallbackHandler)
+        if use_https:
+            if not Path("localhost.pem").is_file() or not Path("key.pem").is_file():
+                create_pem_files()
+            httpd.socket = ssl.wrap_socket(
+                httpd.socket,
+                server_side=True,
+                certfile="localhost.pem",
+                keyfile="key.pem",
+                ssl_version=ssl.PROTOCOL_TLS,
+            )
+
+        httpd.timeout = self.httpd_timeout
+        return httpd
+
+    def get_access_token(self, auth_code):
+        """Exchange an auth code for an access token"""
+        data = {
+            "client_id": self.client_info.client_id,
+            "client_secret": self.client_info.client_secret,
+            "grant_type": "authorization_code",
+            "redirect_uri": self.client_info.callback_url,
+            "code": auth_code,
+        }
+        return requests.post(
+            self.client_info.token_uri, headers=HTTP_HEADERS, data=data
+        )
+
+    def refresh_token(self, refresh_token):
+        data = {
+            "client_id": self.client_info.client_id,
+            "client_secret": self.client_info.client_secret,
+            "grant_type": "refresh_token",
+            "refresh_token": refresh_token,
+        }
+        response = requests.post(
+            self.client_info.token_uri, headers=HTTP_HEADERS, data=data
+        )
+        self.validate_response(response)
+        return safe_json_from_response(response)
+
+    def revoke_token(self, access_token):
```

Comment:
Do we use this anywhere?

Reply:
No. This was present in the `salesforce.py` when I copied things over. Would you like me to remove this (and the references to `revoke_uri` in other places)?

Reply:
I suspect the intent was that we might explicitly revoke the access token once a cci command is done. So it probably makes sense to keep this and add that separately.